Restructure contents and add more details to process #8

damithc opened the pull request:
Fixes #8

damithc commented:
@pyokagan @wkurniawan07 @acjh 

Please have a look (at least at areas that you have more interest in) and give high-level comments. 
The objective is to make this process more detailed but also general so that all our projects can use it as a baseline and define their own process as deviations only.

damithc commented on docs/AdvancedContributorGuidelines.md:
````diff
@@ -0,0 +1,48 @@
+  [4/4] testutil: remove TaskBookBuilder
+  ```
+  
+  > TODO : add a link to the page for generating PR summary
````
@pyokagan can we give here the link to the page you created?

pyokagan replied:
@damithc It's https://pyokagan.github.io/prsummary-generator/

I'm not supporting or maintaining it though, so I think you will need to set up an actual project to maintain the supporting tools for this workflow.

How's reviewable.io, by the way? :-P Will it be worth researching on a better code review system to migrate to, or should we stick with GitHub comments + some of our own tooling?

damithc replied:
@pyokagan 

> I'm not supporting or maintaining it though, so I think you will need to set up an actual project to maintain the supporting tools for this workflow.

I'll use this one for now. We can fork it to oss-generic later.

damithc commented on docs/DefiningLabels.md:
```diff
@@ -0,0 +1,111 @@
+
+Status labels can be applied to PRs as well.
+
+* `s.OnHold` : The work on the PR has been put on hold pending some other event.
```
I'm thinking of trying to operate without status labels.
1. The new 'review status' lowers the need for status tables
2. Status labels often go out of sync with the actual status because some contributors don't have permissions to add labels and others forget to update the label.

The downside is that we cannot filter issues/PRs by status. 

In the meantime we can keep looking for ways to overcome the point 2 above, possibly by automating application of status labels.

pyokagan replied:
> The new 'review status' lowers the need for status tables

I have trouble finding out how I can ask GitHub to just show me "PRs that I have to review" or "PRs I am a reviewer of" (The "PRs I am mentioned in" is useless because I'm mentioned quite a lot of times, and GitHub does not have a way of removing a PR from that list once I'm mentioned even once :-/)

Maybe I'm just bad at searching, but if GitHub doesn't have this feature then the "reviewer status" feature is quite useless and I'd much prefer assigning myself to the PR so that I can "bookmark" and monitor it.

damithc replied:
You are right @pyokagan GitHub filter syntax hasn't caught up with the 'reviewers' features.  Even reviewable.io uses the `assignee` field to find reviewers instead of the `reviewers` field.

damithc replied:
I'll revise the workflow to keep using `assignees` in parallel to `reviewers`. Let's try without status labels for a while to see if that can work.

pyokagan commented on docs/QuestionsIssuesPrs.md:
```diff
@@ -0,0 +1,109 @@
+8. After creating the PR, 
+   1. Wait for CI (i.e. Travis, AppVeyor) to run tests/checks against your PR and report status. 
+      If any errors are reported, fix those problems and push the fixes to the same branch.
+   2. When there are no more CI errors, post a comment to inform the dev team that the PR is ready for review. 
```
Also @mention developers (even those not part on the dev team) who have touched the same area of the code as the PR.

damithc replied:
Unlike normal OSS projects, many of our devs are no longer active so many of the mentions can go unheeded. I'll see if we can work this in somehow.

pyokagan replied:
@damithc 
> many of our devs are no longer active so many of the mentions can go unheeded.

Well, we don't have to *expect* to get a response :-)  If they do respond, though, sometimes old devs can provide valuable insights. So, I'd recommend it since mentioning is a pretty low-cost thing to do.

damithc replied:
@pyokagan 

> I'd recommend it since mentioning is a pretty low-cost thing to do

It _does_ some effort to find the devs to mention, yes? Is there an easy way to do it (e.g. one git command)
Sounds like a job for this bot :-p https://github.com/facebook/mention-bot

pyokagan replied:
@damithc 
> It does some effort to find the devs to mention, yes?

Yes it does require some effort, either by manually browsing the history or installing some git utility scripts. But if you think about it, we (those handling the PRs) are already doing it anyway. e.g. when you pick reviewers like in https://github.com/se-edu/addressbook-level4/pull/202#issuecomment-266179194

This change:
> Also @mention developers (even those not part on the core dev team) who have touched the same area of the code as the PR.

would push this effort to the *contributors*, thus making it easier for the core devs :-)

@damithc 
> Sounds like a job for this bot :-p

Yes, that would help.

pyokagan commented on docs/FormatsAndConventions.md:
```diff
@@ -0,0 +1,62 @@
+
+format: **`IssueTitle #IssueNumber`**
+
+e.g. `Error alert email has very long subject #5958`
```
Hrm, issue titles tend to describe a *symptom* of a problem and thus do not really describe pull requests (the *solution*) well in practice though.

damithc replied:
Good point @pyokagan It's a simplification we use for easier tracing (and also to cut down rubbish PR titles that we are likely to get from newbie contributors such as GSoC hopefuls :-p) 
Going to add the following clarification.

> Duplicating issue title in PR title is for easy tracing between PRs and issues, to compensate for GitHub's lack of strong
> linking between the two.
> Assume there is an invisible prefix in front of the PR title `Fixes issue : ...`

pyokagan commented on docs/MergingPrs.md:
```diff
@@ -0,0 +1,34 @@
+
+**Who can merge?** 
+A PR is to be merged by the author of the PR if the author has commit privileges. 
+Else it should be merged by the main reviewer.  
```
@damithc Since all we do is press the "merge" button, why can't the project manager do this directly? This would shorten turnaround time.

damithc replied:
Yes, in some cases PM can do it (e.g. if no manual testing is needed). I'll add that.

pyokagan commented on docs/OverseeingIssuePrLifeCycle.md:
```diff
@@ -0,0 +1,36 @@
+
+* A PR should have,
+  * **1 or more `reviewers`** (See [Reviewing PRs](ReviewingPrs.md) for more info)
+  * **no more than one `assignee`** : The `assignee` field can be used to indicate the main reviewer when there are 
```
So, yeah, as long as the "reviewers" feature remains not-very-useful, I hope this restriction can be relaxed.

damithc replied:
Noted.

pyokagan commented:
Re: Commits and commit messages, some additional ideas for inspiration: https://github.com/git/git/blob/e05806da9ec4aff8adfed142ab2a2b3b02e33c8c/Documentation/SubmittingPatches#L37-L132 https://github.com/git/git/blob/e05806da9ec4aff8adfed142ab2a2b3b02e33c8c/Documentation/SubmittingPatches#L303-L317

wkurniawan07 commented on README.md:
```diff
@@ -1,94 +1,46 @@
 # A Reference Process for OSS Projects on GitHub
 
-This is an attempt to document similarities in in a number of our OSS projects hosted on GitHub. 
+This is an attempt to document practices common to a number of OSS projects we have on GitHub. 
```
This is one of the reasons why I questioned the difference between `oss-generic` and `nus-oss` even back then.

> ...OSS projects **we** have on GitHub"

"We" here represents the team from NUS-OSS, and in no way we are _the_ group appointed by GitHub to be some kind of ambassador. We may have set up a workflow followed in all our projects, we may welcome all other projects to adopt/adapt (as mentioned soon after), but in no way we are _the_ authority for GitHub's OSS workflow (if it even exists).

damithc replied:
You are right @wkurniawan07 it's not clear who this `we` are. 

1. As you may notice, some of our orgs have `nus` in the name while others don't.  Those meant for 'NUS students only' have `nus` in it. That is because NUS employees are not allowed to use NUS brand to push products/services to third parties as it can expose NUS to legal liabilities. Also, keeping `nus` out makes them more neutral to adopters from peer universities. That is why we have not moved teammates etc. under nus-oss org. They remain independent orgs although mentioned in nus-oss org. Similarly, this org will contain workflows, conventions, tools etc we developed but available for outsiders. There is no need to link any of it to NUS just because most of us (but not all) involved in those projects are in NUS (or was in NUS at some point).

2. I don't think this repo is trying to claim as representing GitHub and we are in no position to impose this process on anyone either. We write these docs in a way that projects who follow that process can use the docs directly instead of duplicating the same content in their own docs. Others can 'take it or leave it'.  The name `oss-generic` sounds general, but it's just an identity. I wanted it to sound neutral and not connected to nus so that an outside project can say 'we follow the oss-generic process as described at {link}'.  It's like 'semantic versioning';  we don't question how come some random guys decided to name their versioning convention as 'semantic'.  

But you are right that the way it is presented can be misleading to readers. Going to revise as follows:

---

# The _OSS-Generic_ Reference Process 

The _OSS-Generic Reference Process_ described in this repo documents common practices among several OSS projects on GitHub, notably, [TEAMMATES](https://github.com/teammates/teammates), [SE-EDU projects](https://github.com/se-edu/main), [PowerPointLabs](https://github.com/powerpointlabs/powerpointlabs).

Other projects are welcome to adopt/adapt these practices if they wish to.

pyokagan replied:
@damithc 

Some of my thoughts, fwiw,

The name of the organization reflects its purpose. For me, the name "oss-generic" leads me to wonder (and worry) about the scope of the project.

Who/What is this project catering to? Is it catering to TEAMMATES, SE-EDU, PowerPointLabs? Or is it catering to the OSS community at large?

Semantic versioning, for example, caters to the OSS community at large. As such, the project has to deal with [116 open issues and 12 open PRs](https://github.com/mojombo/semver) at the time of writing. And that is only for a simple *version number*. Imagine if this project starts having workflow processes and coding standards.

Now, let's say this project caters to the OSS community at large. Let's assume it catches on, and other projects start adopting the coding standards/processes defined in the documents of this project. We are going to start having requests from other projects demanding us to tweak our documents to match their ideal workflow, or complain we change something which is at odds with their workflow. This means that *our* projects (se-edu, teammates etc.) will lose some degree of autonomy. 

Now, workflow processes contain a lot of human factors inside. For example, the Linux and Git projects still use emails since their contributors (at least, the ones that actually matter) prefer to use Email. This is at odds with what the general OSS communities want, which is to use GitHub. So, if our projects' workflow processes become beholden to the OSS community at large, we may start to have friction between what *our* contributors want and what the general OSS community wants. I would personally prefer our projects to have more flexibility and autonomy.

We could, of course, put our foot down and say "No, we are mainly serving our own projects", but then we will then get lots of unnecessary drama ;-)

Of course, all these problems won't happen if this project does not catch on with the wider OSS community, but then this would also not happen if the org was named something else.

So, I think that this project should make it very clear that it exists to *only serve* the TEAMMATES, SE-EDU, PowerPointLabs projects. This should be shown through its name (which should not imply that it is primarily catering to the OSS community at large). Of course, the wider OSS community is free to adopt them (as long as they credit us, a proper license will solve that), but I think that should not be the *main* purpose of this project.

damithc replied:
BTW, I'm open to change `OSS-Generic` to a more suitable name (as long as it doesn't have `GitHub` or `NUS` in it). It's just that I couldn't think of a better one at the time.

damithc replied:
@pyokagan 

>The name of the organization reflects its purpose. For me, the name "oss-generic" leads me to wonder (and worry) about the scope of the project.
Who/What is this project catering to? Is it catering to TEAMMATES, SE-EDU, PowerPointLabs? Or is it catering to the OSS community at large?

This is a valid point. My response is, _the project currently caters for a selected set of projects but has the ambition to cater for a wider community, if the wider community accepts it_. 

1. In the more likely case that no one else other than us take note of this repo, there is no problem. 
2. In the case it catches on and we are forced to cater for the wider community's demands at the expense of our own projects, that's a _happy problem_ isn't it? It would be great achievement for all of us involved if the wider community adopts it. So shouldn't we keep the door open for that possibility (however small it is) rather than close it for sure? At the moment, keeping that door open doesn't cost us anything. 

But I kind of feel that it's a bit arrogant(?) to name it OSS-Generic. As I mentioned above, I'm open for a better name that doesn't make sweeping claims but also not NUS specific. The problem is, NUS is the only common thing among our projects. I can't very well name it `Process for projects done by Damith's students` :-D

wkurniawan07 replied:
@damithc 

> ... has the ambition to cater for a wider community, if the wider community accepts it.

I have no qualms about being ambitious - in fact I'm all for it - my concern here is that a large part of this repo's content (those that I can identify: naming conventions, coding standards, community structure) is very specific to our NUS-OSS projects, and thus its validity for the larger OSS community is questionable. If we can make the contents of this repo really generic, e.g. "Your project should have a well-defined convention for issue titles, PRs, etc." (generic) instead of "Here are the conventions you should follow" (specific), then we can justify the generic-ness of the process better. Also, even generic problems like "Why have conventions for ...?" can be subjected to debates, never mind when the conventions are specific.

Where does this lead to? The only way I see out of this is to maintain another process repo in `nus-oss` that contains our specific guidelines. It becomes something like this:
```
oss-generic: can be used by the OSS community at large
    ^
    |
    | inherits
    |
nus-oss: to be used by NUS-OSS projects, e.g. TM, SE, PL; contains specific guidelines
    ^
    |
    | inherits
    |
individual projects
```
This way, the individual projects stay independent from the name "NUS" (unless following `nus-oss` guidelines already breaches it), we have a repo (`nus-oss/process`) containing processes common to all our projects, and we have a repo for generic processes that others can use, with no pollution (for the lack of better word) from our projects-specific guidelines.

damithc replied:
@wkurniawan07 
If we say something like `prefix PR number with your matric number`, that's specific to NUS. None of the guidelines we have are specific to NUS. I tried to exclude any NUS specific things on purpose. If I missed any, do point out so that I can get rid of it. I think anyone running an OSS project on GitHub and has no preconceived idea of how to run it (e.g. if that's their first OSS project) can follow our process exactly as its prescribed and they'll do just fine. Any thoughts?

The point you raise, which is valid, is actually about the level of concreteness, not about the level of specificity. Our guidelines are very concrete (e.g. `copy paste issue title as PR title`) rather than abstract (e.g. `PR title should be clearly linked to issue title`). The concreteness, while being very much influenced by our personal preferences, also makes them immediately useful. It took a lot of trial-and-error for us to figure out those concrete details that works reasonably well. Some others (obviously not all) may be able to simply follow our final version of the process, instead of having to figure out themselves all those details. For example, PhD students who are developing OSS projects as part of their research can follow this process irrespective of whether they are in NUS or not. Those guys are unlikely to have lot of experience running OSS projects themselves, so they'll be starting with a clean slate. By simply copying our process, they can have a sizeable head start instead of starting from zero.

As I write this answer, I also realized one possible way we can narrow down the scope. Instead of claiming that it fits all OSS projects, we can say that it can help those who are new to running OSS projects, who can use this as a 'starter process' that they can adopt partially or fully at the start and then evolve as they see fit as they gain more experience and start developing their own ideas of how to do OSS things.

pyokagan replied:
@damithc 
> As I mentioned above, I'm open for a better name that doesn't make sweeping claims but also not NUS specific.

Maybe this could be part of SE-EDU? Since besides documenting what TEAMMATES, SE-EDU, PowerPointLabs already do, it also serves to *educate* students/developers who would like to contribute/set up an OSS project? After all, some of these stuff (e.g. coding standards) are applicable to both OSS and non-OSS projects.

damithc replied:
@pyokagan 

> Maybe this could be part of SE-EDU?

That's another idea with merit. It's not an exact fit, but there is significant overlap between the two. I'm also open for moving it to NUS-OSS or forking it there as @wkurniawan07 suggested. 

Anyway, the revised README claims this repo documents common practices between a number of specific projects. I'll remove the line that says others are welcome to adopt it. After that, it's just a repo in the public domain that happens to have the org name `oss-generic` for some unknown reason. It contains exactly what it claims to contain. I don't think anybody will object to it or make demands for changes. So, for all practical purposes let's treat it as process for our own projects. We can do that without specifically mentioning NUS, the same way we try to keep our other projects (e.g. SE-EDU) independent from NUS. It's important to do so to preserve their freedom to chart their own path in future, instead of being tied to a specific institution. All of these projects are meant to grow beyond NUS, although they are currently concentrated in NUS. The process documented in OSS-Generic is not mature yet, so I don't think anybody would want to adopt it anyway. We can revisit this topic perhaps one year down the line and see if we have something we can offer to the rest of the world.

wkurniawan07 commented on README.md:
```diff
@@ -1,94 +1,46 @@
+<br>
 
-# Coding Standards
+**Contributing to this repo**: We welcome contributions. Just follow the instructions in the above links to submit 
```
I feel like this can be a heading on its own, because contributing is a section separate from "the list of documents we provide". Of course, feel free to disagree.

damithc replied:
Agreed. For the moment, I'll make it similar to other headings (keeping all headings at same level to keep the page structure simple), and make all of them proper headings.

wkurniawan07 commented on README.md:
```diff
@@ -1,94 +1,46 @@
-
-We welcome contributions. Just follow the process described above to submit your contributions as a PR.
\ No newline at end of file
+[gh]: https://github.com/
```
I don't think this is necessary; there are no anchors to this anywhere on the file.

damithc replied:
Good catch. Forgot to remove. Will remove now.

wkurniawan07 commented on README.md:
```diff
@@ -1,94 +1,46 @@
+* [**Asking a Question**](docs/QuestionsIssuesPrs.md#asking-a-question)
+* [**Posting an Issue**](docs/QuestionsIssuesPrs.md#posting-an-issue)
+* [**Submitting a PR**](docs/QuestionsIssuesPrs.md#submitting-a-pr) 
+* [**Advanced Contributor Guidelines**](docs/AdvancedContributorGuidelines.md) : Additional guidelines more 
```
I personally prefer no whitespace before `:`, but it's personal so it does not need to be followed. The more important thing is, there are variations of `abc: def` and `abc : def` throughout this repo.

damithc replied:
Good point. I'll remove the space.

wkurniawan07 commented on README.md:
```diff
@@ -1,94 +1,46 @@
+* [**Coding Standards**](docs/CodingStandards.md)
+* [**'How to' Guides**](docs/HowToGuides.md) : Detailed help on how to perform some common tasks such as 
+  editing a commit message.
+* [**Formats and Conventions**](docs/FormatsAndConventions.md) : Fromats and conventions for PRs, Issues, Files, etc.
```
~~Fromats~~ Formats

damithc replied:
Will fix.

wkurniawan07 commented on docs/AdviceForOssBeginners.md:
```diff
@@ -0,0 +1,26 @@
+
+## Do your homework
+
+Use Google, read trouble shooting guides, search issue tracker (including closed issues)
```
This one point taken from the TEAMMATES contributor's orientation guide sums up my feelings when new contributors ask help for the most trivial of tasks, and they really should know:

> Contributing to an OSS project requires you to figure out things on your own when you can, and seek help from the right resource _(Google, StackOverflow, troubleshooting guides, issue tracker, etc.)_ when you cannot. _You will not make a good impression to the dev team members if you ask questions which are already answered in the mentioned resources._

Sentences in italic are my own addition. I apologize if I sound too harsh, but the idea is there.

damithc replied:
Will add your version.

wkurniawan07 commented on docs/CodingStandards.md:
```diff
@@ -0,0 +1,12 @@
+* C#
+* CSS
+* **Java**: [[Style guide](https://oss-generic.github.io/process/codingstandards/coding-standards-java.html)]
+  [Best practices]
```
I think the wrapping `[...][Best practices]` is unnecessary. Similarly for the HTML one.

damithc replied:
Removed.

wkurniawan07 commented on docs/AdviceForOssBeginners.md:
```diff
@@ -0,0 +1,26 @@
+# Advice for OSS Beginners
+
+//TODO: flesh out these points
```
Is this TODO still valid? If only two points are not fleshed out, the TODO should be moved there instead.
If necessary I can come up with some good descriptions for those points.

damithc replied:
Will remove.

damithc commented:
On second thoughts, may be we should retain status labels? They are useful to identify the state of an issue from the issue list without having to go into the issue page. Reviewer status is not shown in the issue list. @pyokagan @wkurniawan07

damithc commented:
@wkurniawan07 , let's wrap up this soon so that you can update TEAMMATES docs accordingly.

Any input on the status labels thing? Shall I put back instructions to use them (as we are doing now)?

acjh commented on docs/QuestionsIssuesPrs.md:
```diff
@@ -0,0 +1,109 @@
+   * Commit code at suitable points. Follow the [conventions for commit messages](FormatsAndConventions.md#commit).
+
+7. When you think you are done with fix, create a new PR. Here are the steps: 
+   1. **Update a PR with upstream changes**: There is a chance that While you were write code for the PR, 
```
Typo for unnecessarily capitalized 'w' in `While`.

acjh commented on docs/QuestionsIssuesPrs.md:
```diff
@@ -0,0 +1,109 @@
+    with the `d.Contributors` label.
+  * As harder issues take longer to finish, it is prudent to post a message in the issue to let others know that you are 
+    working on an issue.
+  * As you submit more (and bigger)PRs, incrementally adopt 
```
Missing space between `(and bigger)` and `PRs`.

acjh commented on docs/ReviewingPrs.md:
```diff
@@ -0,0 +1,54 @@
+ * PM
+
+Additional reviewers may be used if the PR involves multiple areas.
+Junior members can get their PR peer reviewed by another junior member before submitting to the projects formal 
```
Missing apostrophe (`'`) in `projects`.

acjh commented on docs/ReviewingPrs.md:
```diff
@@ -0,0 +1,54 @@
+
+A reviewer must do the following steps.
+
+1. Ensure the the following at the start of the review. If any of them are missing, ask the dev to fix.
```
Typo for repeated `the`.

acjh commented on docs/ReviewingPrs.md:
```diff
@@ -0,0 +1,54 @@
+1. Ensure the the following at the start of the review. If any of them are missing, ask the dev to fix.
+  * PR name follows the [PR title convention](FormatsAndConventions.md#pr). 
+  * PR description has the correct `Fixes #...` reference.
+  * The CI build is successful or any failures are justifiable (e.g. false positives from static analysis tools) 
```
Missing a full stop.

acjh commented on docs/ReviewingPrs.md:
```diff
@@ -0,0 +1,54 @@
+    1. **Tests**:  Almost all code changes to functional code should have changes to test code.
+    1. **User docs** e.g. help pages.
+    1. **Dev docs** e.g. design diagrams
+  * Coding style and best practices are followed (some of these are not detected by static analysis tools)
```
Missing a full stop.

acjh commented on docs/ReviewingPrs.md:
```diff
@@ -0,0 +1,54 @@
+  * The PR does not contain unrelated changes. 
+      e.g. unnecessary formatting changes or commits from other branches.
+
+1. Use [GitHub's review feature] to add comments.
```
Missing a hyperlink?

acjh commented on docs/ReviewingPrs.md:
```diff
@@ -0,0 +1,54 @@
+  append `?w=1` to url of the `/files` page of the pull request (the "Files changed" tab)
+
+* **If the PR is too big**, explore the possibility of breaking the PR into multiple self-containing steps
+  as explained in [Creating commits in big PRs](AdvancedContributorGuidelines.md#creating-commits-in-big-prs)here. 
```
- Use 'self-contained' (33,600,000 results) instead of `self-containing` (122,000 results).
- Missing space before `here`.

acjh commented on docs/ReviewingPrs.md:
```diff
@@ -0,0 +1,54 @@
+    1. **Code**
+    1. **Comments** e.g. header comments
+    1. **Tests**:  Almost all code changes to functional code should have changes to test code.
+    1. **User docs** e.g. help pages.
```
Avoid full stop for consistency.

damithc commented:
Thanks for the bug reports @acjh. Fixed.

damithc commented:
Two updates:

1. Introduced a `Code Quality Reviewer` role to the PR review workflow. Instead of PM, it will be the code quality reviewer who will do the final approval. PM will also be the code quality reviewer in most cases, but when needed (e.g. if the PM has no expertise in the project code), the role can be given to someone else. 67529b66dd4396824b96c427381682a04eb26a89
2. Added back the status labels to the workflow. I think it is premature to drop them at this point because GitHub's 'review request' feature is still half baked. ffe945260a1df73e892793e1a840c71f576026e4

acjh commented on docs/ReviewingPrs.md:
```diff
@@ -0,0 +1,63 @@
+# Reviewing PRs
+
+A typical PR is reviewed by,
+ - [x] one _**main reviewer**_ : Usually, a lead member. Reviews _all_ aspects of the PR, including code quality. 
```
Remove space after colon?

damithc replied:
Removed the space before the colon, to be consistent with the other places.

acjh commented on docs/ReviewingPrs.md:
```diff
@@ -0,0 +1,63 @@
+ - [ ] [Optional] _**supplementary reviewers**_ : Core members can be used as additional reviewers if the PR involves
+       multiple areas. 
+
+> Junior members can get another junior member to act as a _**peer reviewer**_ for their PRS before submitting 
```
'PRs' instead of `PRS`. Or singular 'PR' to be consistent within this section.

damithc replied:
Fixed.

damithc commented:
Just realised that GitHub preserves comments on an commit even if the commit was later overwritten by a force push. For example, I fixed the two typos @acjh mentioned in the latest review, squashed the 'fix typo' commit to the offending commit that introduced the typos and contains the comments, and force pushed. @acjh comments are still preserved and the code in the comment context still shows the typo although the typo no longer exist in the branch anywhere. That means rewriting past commits does not destroy the discussion history.

acjh commented:
I believe the comments are referenced directly by the PR. If you click the link to the file or `View changes`, GitHub says it couldn't find those commits.

acjh commented:
I think it's meant for the discussion history **not** to be indirectly destroyed by a force-push, which is good.

wkurniawan07 commented on README.md:
```diff
@@ -1,94 +1,40 @@
+  on overseeing issue/PR life cycles.
+* [**Managing Releases**](docs/ManagingReleases.md): Instructions for managing releases, milestones, 
+  and project schedule.
+* [**Defining Labels**](docs/DefiningLabels.md): Instructions for merging a PR.
```
The description is wrong :p

damithc replied:
Will fix

wkurniawan07 commented on docs/AdviceForOssBeginners.md:
```diff
@@ -0,0 +1,36 @@
+## Follow others
+
+> OSS projects involve many persons contributing in small measures. If each person did things their own way, the project
+> will deteriorate into a mess very quickly. That is why being consistent is very important in OSS projects.
```
I don't think this part needs to be block-quoted. It's good as normal paragraph.

damithc replied:
I use block quotes for additional info that can be skipped. e.g. rationales, extra tips. This paragraph fits that category. What do you think?

wkurniawan07 replied:
I use block quotes for that purpose as well, but to me this does not look "skippable". Of course, to each his own :p

wkurniawan07 commented on docs/AdviceForOssBeginners.md:
```diff
@@ -0,0 +1,36 @@
+
+## Do your homework
+
+Use Google, read trouble shooting guides, search issue tracker (including closed issues).
```
This sentence is already part of the "seek help from the right resource" so it may be redundant/repetitive.

damithc replied:
Will fix

wkurniawan07 commented on docs/DefiningLabels.md:
```diff
@@ -0,0 +1,123 @@
+
+* Open issues
+    * No status: New issue yet to be triaged. 
+    * `s.Ongoing` : The issue is being worked on.
```
I believe we use the `s.Ongoing` label anymore.

damithc replied:
Will remove

wkurniawan07 commented on docs/DefiningLabels.md:
```diff
@@ -0,0 +1,123 @@
+    * `s.MergeApproved` : Both main reviewer and the code quality reviewer has approved the merge. PR can be merged.
+    * `s.OnHold` : The work on the PR has been put on hold pending some other event.
+* Closed issue/PR
+    * `s.Discarded` : The PR was abandoned. 
```
Actually, what's the purpose of this label? If all closed-without-merging PRs are labelled with this, it is as good as filtering with the correct keywords. If this kind of label is necessary, we need to have correct variations such as `s.Discarded`, `s.Invalid`, `s.Duplicate`, and so on.

damithc replied:
This was introduced to easily identify discarded PRs through HubTurbo UI (Hubturbo doesn't show the merge status). Will remove.

wkurniawan07 commented on docs/DefiningLabels.md:
```diff
@@ -0,0 +1,123 @@
+    * `s.Ongoing` : The PR is being worked on.
+    * `s.ToMerge`: Main reviewer approved the changes. 
+    * `s.MergeApproved` : Both main reviewer and the code quality reviewer has approved the merge. PR can be merged.
+    * `s.OnHold` : The work on the PR has been put on hold pending some other event.
```
This one could be applied to issues as well.

wkurniawan07 replied:
Also, on this section, the space before `:` needs to be removed.

damithc replied:
Will fix both

wkurniawan07 commented on docs/DefiningLabels.md:
```diff
@@ -0,0 +1,123 @@
+|`a-Scalability`    |Related to behavior at increasing loads
+|`a-Security`       |Protection from security threats
+|`a-Testing`        |Testing efficiency and robustness (as opposed to testing a specific feature)
+|`a-UIX`            |User interface, User experience, Responsiveness
```
1. May want to add that `These are aspects that are common to most software projects, open source or not.`
2. `a-Persistence` is TEAMMATES-specific
3. `a-Docs` may want to remove the "Website" reference because not all projects use that, or make it "User docs, e.g product website"

damithc replied:
1. Not sure if it adds any value. Will add `Some examples:` to indicate this is not an exhaustive list.
2. Will rephrase to make it not specific to TEAMMATES. Some sort of persistence is there in most projects.
3. will do.

wkurniawan07 commented on docs/FormatsAndConventions.md:
```diff
@@ -0,0 +1,65 @@
+
+## Branch
+
+format: **`issueNumber-some-keywords-from-issue-title`**
```
1. ~~f~~Format (also for other sections).
2. Also we may want to justify why we have this kind of convention. We used to have `Issue1234` before and moved to this convention for a reason. Also for the merge commit message (PR name already explained).

damithc replied:
1. Will fix
2. Will add

wkurniawan07 commented on docs/HowToGuides.md:
```diff
@@ -0,0 +1,28 @@
+
+1. Sync your branch with the `master` branch. This can be done in two ways:
+   1. Merge `master` to your branch (this option is recommended for new contributors).
+   2. Rebase the PR branch on the `master` branch (this option is recommended for experienced contributors). 
```
1. Use `1` for consistence.
1. ~~this option is~~ does not seem to value-add.

damithc replied:
Will fix

wkurniawan07 commented on docs/ManagingReleases.md:
```diff
@@ -0,0 +1,66 @@
+    * Project lead needs to periodically check the issue tracker for inactive PRs and remind the relevant 
+      persons to resume work.
+    * If a PR is inactive even after a reminder, it can be closed (and the issue unassigned). 
+    * TODO: automate the reminders.
```
This seems like it is for our projects, so it should be removed.
Otherwise, it should be made generic like "Some third-party services or bots may be able to automate those reminders".

damithc replied:
Will remove

wkurniawan07 commented on docs/OverseeingIssuePrLifeCycle.md:
```diff
@@ -0,0 +1,39 @@
+## Triaging Issues
+
+Issue tracker doubles as the message board. Therefore, it can have two types of issues:
+* _Support Requests_: (i.e. announcements, questions, support requests, self-intros etc.) 
```
The brackets should not be necessary.

damithc replied:
Will remove.

wkurniawan07 commented on docs/OverseeingIssuePrLifeCycle.md:
```diff
@@ -0,0 +1,39 @@
+  * If you can, assign at least one `reviewer`. If you don't, team lead will assign reviewers for the PR.
+  
+  > To find suitable reviewers, you can use GitHub's _blame_ feature. 
+  > **Currently active** Developers who have touched the same code before are potential reviewers.
```
Can also mention "Area Leads" of the project.

damithc replied:
Will do

wkurniawan07 commented on docs/QuestionsIssuesPrs.md:
```diff
@@ -0,0 +1,109 @@
+## Posting an Issue
+
+1. Do a search in the issue tracker to ensure the same issue has not been posted before.
+2. Follow the [naming convention](FormatsAndConventions.md#issue) for the issue title. 
```
The numbering on this document should also use all `1`s.

damithc replied:
Will change

wkurniawan07 commented on docs/QuestionsIssuesPrs.md:
```diff
@@ -0,0 +1,109 @@
+
+1. Read any instructions the project has for new contributors. Some projects require you to read some documents
+   and/or do some orientation activities before you start contributing. 
+   For example, [TEAMMATES](../../../teammates/teammates) has a 'Contributor Orientation Guide'.
```
1. Using relative may work, but it should be better with absolute.
2. Also, may want to link to TEAMMATES' contributor orientation guide.

damithc replied:
Will remove it entirely.

wkurniawan07 commented:
Also my two cents, fwiw:

1. On status labels for review: yes, we can't dispose of them just yet because GitHub's review feature is not quite there yet for all our usages. Having said that, we can still use what they have to offer, which we have been doing.
2. On force-pushing and comments: I recalled some time back that it is indeed the case, but comments that are _attached to commits_ (not attached to a line of change) will be removed by force-push.

damithc commented:
Updated based on @wkurniawan07's review.

The other notable update: issue posts were previously divided into `Change Requests` and `Support Requests`. Renamed the latter to `Misc Posts` to better represent wide variety of posts it covers.

acjh commented on docs/OverseeingIssuePrLifeCycle.md:
```diff
@@ -25,8 +25,9 @@ When an issue comes in,
   
-  > To find suitable reviewers, you can use GitHub's _blame_ feature. 
-  > **Currently active** Developers who have touched the same code before are potential reviewers.
+  > GitHub's _blame_ feature (or use a third party service such as _mentionbot_)can help to find suitable reviewers. 
```
- The word `use` is unnecessary.
- Missing hyphen in `third party`.
- Missing space before `can`.

damithc replied:
Fixed

wkurniawan07 commented on docs/AdviceForOssBeginners.md:
```diff
@@ -0,0 +1,38 @@
+* If the setup instructions mention a specific version of a software to use, **use that exact version** 
+  instead of the latest version.
+* Do not skip any steps because you think that step is "not important". 
+* Do not assume current project does things the same way things were one in that _other_ project you are familiar with.
```
"that _other_ project" seems like we're trying to imply something. "other projects" is more tone-neutral.

damithc replied:
Will fix

wkurniawan07 commented on docs/FormatsAndConventions.md:
```diff
@@ -0,0 +1,74 @@
+Format: **`[#IssueNumber] Issue Title (#PrNumber)`** <br>
+e.g. `[#5958] Error alert email has very long subject (#6580)`
+
+> Rationale: This format allows easily traceability between a merge commit, the issue it fixes, and the PR that fixed it. 
```
eas~~il~~y traceability

damithc replied:
Will fix

wkurniawan07 commented on docs/AdvancedContributorGuidelines.md:
````diff
@@ -0,0 +1,49 @@
+  [4/4] testutil: remove TaskBookBuilder
+  ```
+  
+  > You can use [this page](https://pyokagan.github.io/prsummary-generator/) (created by @pyokagan)
````
GitHub does not automatically attach anchor to usernames in docs, so Paul's username here is not automatically linked. Unless it was never intended to begin with?

damithc replied:
Wasn't intended to. Will add link.

wkurniawan07 commented on docs/AdvancedContributorGuidelines.md:
```diff
@@ -0,0 +1,49 @@
+## Updating a PR in response to a review
+  
+* If the `master` has progressed beyond the root of the PR branch, rebase the PR branch 
+  (instead of merging master to PR branch). This will keep the PR branch free of merge commits.
```
`master` should be `coded` here.

damithc replied:
Will fix

wkurniawan07 commented on docs/MergingPrs.md:
```diff
@@ -0,0 +1,41 @@
+  1. If the commit history is clean (this should be the case for PRs done by experienced contributors who follow 
+     the [advanced contributor guidelines](docs/AdvancedContributorGuidelines.md)), <br>
+     Merge using the `Create a merge commit` option.
+  2. Else, merge using the `Squash and merge` option.
```
Looks like I missed this. Change to `1`.

damithc replied:
Will change

wkurniawan07 commented on docs/QuestionsIssuesPrs.md:
```diff
@@ -0,0 +1,108 @@
+   1. **Update a PR with upstream changes**: There is a chance that while you were write code for the PR, 
+      the master branch of the _upstream repo_ (i.e. the original repo you forked from) has been updated with new code. 
+      In that case you need to update your fork and your clone with that new code so that the only difference between 
+      the code in your branch and the upstream master branch is the code you wrote.  <br>
```
Again, `master` are not `coded` in these sentences.

damithc replied:
Will update

wkurniawan07 commented on docs/HowToGuides.md:
```diff
@@ -0,0 +1,28 @@
+1. Update the `master` branch of your local clone. 
+[Here](https://help.github.com/articles/syncing-a-fork/) are the instructions.
+
+1. Push your master branch to your fork (if you are creating PRs using a fork).
```
Again, `master` is not `coded` here.

damithc replied:
Will change

damithc commented:
Fixed bugs reported by @wkurniawan07 in the recent review.

pyokagan commented on docs/AdvancedContributorGuidelines.md:
```diff
@@ -0,0 +1,49 @@
+* Follow issue priorities. Prefer to take higher priority issues over lower priority issues.
+* Favor issues related to certain features/aspects/techs to deepen expertise in those areas. 
+ 
+## Creating commits in big PRs
```
May wish to define what is meant by "big".

For me, a "big PR" is when the number of logical changes it makes becomes too complex. (And we then need to define what a "logical change" is :-) ) An estimation I make is when the overall diff of a PR starts to approach around 100 lines (both additions and deletions).

Also I think, "breaking commits up for big PRs" is a hard requirement, non-advanced contributors are not exempt from it. So, I'd advice contributors unfamiliar with Git to either 1. Learn how to use Git first, or 2. avoid taking complex issues that require complex PRs to solve.

damithc replied:
@pyokagan I agree. We should make it a hard requirement. 

It's hard to define 'big' even in LoC because sometimes a trivial change (an invasive refactoring) can touch a lot of lines. I think what we are looking for is PRs that are too big to review in one-shot (i.e. a 15 minute window); needs to be reviewed incrementally instead.

Perhaps it's the reviewer, realizing the PR is too big, who should ask the dev to refactor the commits to make it a `multi-step PR`? I also suspect, given devs are not very experienced, only in retrospect they can figure out the correct sequence of steps. So creating commits to reflect the steps will be done after the whole change is done anyway. Senior members (PM, project lead, etc.) who monitor incoming PRs can also step in to intercept big PRs and ask the dev to update it to multi-step format.

pyokagan replied:
@damithc 
> It's hard to define 'big' even in LoC because sometimes a trivial change (an invasive refactoring) can touch a lot of lines.

The thing is that effective reviewing of PRs either require that we either:

1. Eyeball each line of the diff line by line. (Skipping lines means potential bugs, or even security issues if the author tries to sneak a backdoor in. Yes, that happens.), or

2. Perform the same steps as the author (usually involves automated tools e.g. eclipse refactoring) and then compare (diff) our results.

So yes, I think LOC does matter. Too big a LOC count means that we can't use method (1) for reviewing, and if it can't be reviewed using method (2), then the PR author must absolutely break down the PR into smaller steps with smaller LOC.

> Perhaps it's the reviewer, realizing the PR is too big, who should ask the dev to refactor the commits to make it a multi-step PR?

Yes, that usually happens, but the point of putting this down in the documentation is so that contributors are aware of it. e.g. if contributors want to dump a gigantic PR with large LOC count without proper commit organization, they better have an explanation ready for why it is that way.

damithc replied:
@pyokagan Agreed. I'll update docs as a separate PR.

acjh commented on docs/OverseeingIssuePrLifeCycle.md:
```diff
@@ -18,18 +18,21 @@ When an issue comes in,
-    multiple reviewers and it is not clear who is the main reviewer. Otherwise it can be left empty.
+  * **all reviewers should be selected as `assignees`**
+  
+  > Duplicating `reviewers` as `assignees` is needed because GitHub filter syntax cannot filter by reviewers yet.
```
Now possible as of https://github.com/blog/2306-filter-pull-request-reviews-and-review-requests :)

damithc replied:
@acjh Is the feature working for you? In my case, it shows 0 open and 0 closed review requests.

acjh replied:
It appears to be working: https://github.com/oss-generic/process/pulls?q=reviewed-by%3Adamithc

damithc replied:
Apparently, the filter works. It's this button (which I expected to be same as the filter) that doesn't work. Anyway, i'll create an issue to update this doc.

![image](https://cloud.githubusercontent.com/assets/1673303/22359698/781f47de-e484-11e6-8b23-3c019c8924ad.png)
